app/flatpak-builtins-repo.c

```diff
@@ -205,14 +205,109 @@ print_metadata (GVariant   *meta,
     }
 }
 
+static gboolean
+collect_tree_size (GLnxDirFdIterator *iter,
+                   guint64 *count,
+                   guint64 *actual,
+                   guint64 *total,
+                   GError **error)
+{
+  while (TRUE)
+    {
+      struct dirent *dent;
+
+      if (!glnx_dirfd_iterator_next_dent_ensure_dtype (iter, &dent, NULL, NULL))
+        break;
+
+      if (dent == NULL)
+        break;
+
+      else if (dent->d_type == DT_DIR)
+        {
+          g_auto(GLnxDirFdIterator) iter2 = { 0 };
+
+          if (!glnx_dirfd_iterator_init_at (iter->fd, dent->d_name, FALSE, &iter2, error))
+            return FALSE;
+
+          if (!collect_tree_size (&iter2, count, actual, total, error))
+            return FALSE;
+        }
+      else if (dent->d_type == DT_REG)
+        {
+          struct stat stat;
+
+          if (!glnx_fstatat (iter->fd, dent->d_name, &stat, 0, error))
+            return FALSE;
+
+          if (count)
+            *count += 1;
+          if (actual)
+            *actual += stat.st_size;
+          if (total)
+            *total += stat.st_size * stat.st_nlink;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+collect_repo_size (OstreeRepo *repo,
+                   guint64 *count,
+                   guint64 *actual,
+                   guint64 *total,
+                   GError **error)
+{
+  g_auto(GLnxDirFdIterator) iter = { 0 };
+
+  if (!glnx_dirfd_iterator_init_at (ostree_repo_get_dfd (repo),
+                                    "objects",
+                                    FALSE,
+                                    &iter,
+                                    error))
+    return FALSE; 
+
+  return collect_tree_size (&iter, count, actual, total, error);
+}
+
+static gboolean
+print_size (OstreeRepo *repo,
+            GError **error)
+{
+  guint64 count;
+  guint64 actual;
+  guint64 total;
+  g_autofree char *actual_formatted = NULL;
+  g_autofree char *total_formatted = NULL;
+
+  count = 0;
+  actual = 0;
+  total = 0;
+
+  if (!collect_repo_size (repo, &count, &actual, &total, error))
+    return FALSE;
+
+  actual_formatted = g_format_size (actual);
+  total_formatted = g_format_size (total);
+
+  g_print ("Objects in repo: %ld\n"
+           "Size: %s\n"
+           "Unshared size: %s\n",
+           count, actual_formatted, total_formatted);
```

Comment:
This still prints both size and unshared size, and i'm not sure its useful, or how understandable this is.

Reply:
yeah, I have my doubts as well. will just strip this out, I think. I am still curious to learn how much sharing there is between refs and commits on a repository the size of flathub, but maybe this is not the way to do it.

Reply:
The upstream flathub repo will not have any hardlinks though. To find the "total" size you would have to loop over every ref and add up the size of all the files in each commit. (Also, how do you treat parent commits? They are likely to share a lot...)